### PR TITLE
Fix Build graph from image

### PR DESCRIPTION
**Pull request by andromeda-224**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
- Modified to add only one file filter for multiple extensions in the FileInputPane
- Allow multiple images to be selected (and validate semicolon delimited string of paths)
- Placement of multiple images on the graph

### Why Should This Be In Core?
Otherwise Build Graph -> From Image cannot filter/load multiple images of image types properly.

### Benefits
To be able to build a graph by loading multiple images files.

### Possible Drawbacks
Not able to build a graph by loading multiple images files.

### Verification Process
1. Add new graph.
2. Select Experimental -> Build Graph -> From Image...
3. Select multiple images.

Verify that images appear on the graph with no exceptions or warnings.

### Applicable Issues
#2146

### CODE REVIEW

**Review by Quasar985 — APPROVED** [review]

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import au.gov.asd.tac.constellation.plugins.parameters.types.FileParameterType.FileParameterValue;
 import au.gov.asd.tac.constellation.utilities.gui.filechooser.FileChooser;
 import java.io.File;
+import java.nio.file.Paths;
```
Comment (Delphinus8821) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java:29]: Remove unused import

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java [hunk 2]
```diff
@@ -292,21 +293,26 @@ private FileChooserBuilder getFileChooser(final PluginParameter<FileParameterVal
...
         if (extensionFilter != null) {
-            for (final String extension : extensionFilter.getExtensions()) {
-                // Add a file filter for all registered exportable file types.
-                fileChooserBuilder = fileChooserBuilder.addFileFilter(new FileFilter() {
-                    @Override
-                    public boolean accept(final File file) {
-                        final String name = file.getName();
-                        return (file.isFile() && StringUtils.endsWithIgnoreCase(name, extension)) || file.isDirectory();
+            // Add a file filter for all registered exportable file types.
+            List<String> extensions = extensionFilter.getExtensions();
```
Comment (Delphinus8821) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java:297]: final

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java [hunk 3]
```diff
@@ -292,21 +293,26 @@ private FileChooserBuilder getFileChooser(final PluginParameter<FileParameterVal
...
+            
+            fileChooserBuilder = fileChooserBuilder.addFileFilter(new FileFilter() {
+                @Override
+                public boolean accept(final File file) {
+                    final String name = file.getName();
+                    String ext = "";
+                    int dotIndex = name.lastIndexOf('.');
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/gui/FileInputPane.java:304]: final. `ext` could also be made final by converting that declaration to a ternary statement

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java [hunk 4]
```diff
@@ -383,7 +383,16 @@ public boolean isWarningOverwriteUsed() {
         @Override
         public String validateString(final String s) {
             final File validationFile = new File(s);
-            return (validationFile.isDirectory() || (validationFile.getParentFile() != null && validationFile.getParentFile().exists())) ? null : "The specified file path doe not contain valid directories";
+            final String[] names = s.split(SeparatorConstants.SEMICOLON);
+            if (names.length > 1) {
+                if ((this.kind.equals(FileParameterKind.OPEN_MULTIPLE) && files.size() != names.length)) {
+                return "Wrong number of files";
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/FileParameterType.java:389]: fix formatting here

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 5]
```diff
@@ -134,125 +143,134 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+                    final int rgb = image.getRGB(x, y);
+                    final int a = (rgb >> 24) & 0xff;
+                    final int r = (rgb >> 16) & 0xff;
+                    final int g = (rgb >> 8) & 0xff;
+                    final int b = rgb & 0xff;
+
+                    // Generate Z using grayscale.
+                    final float gray = 0;
+
+                    final int vxId = vertexIds[frame][x + prevWidth][y] = graph.addVertex();
```
Comment (OrionsGuardian) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:221]: The logic here can be simplified to just use a 2d array:
`final int[][] vertexIds = new int[w][h];`
Further down in the code, the `frame` variable is only used in the vertex identifier, so is not needed as part of the array structure. 
Also, the `prefWidth` should only be used in setting the x-offset the x,y,z co-ordinates of each vertex. 
The size of the array should simply match the size of the image.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 6]
```diff
@@ -134,125 +143,134 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+        for (final BufferedImage image : images) {
+            // Currently, multiple images will be added on top of the prev one
+            // Uncomment the 3 lines below to add image to the right of prev one
+            if (w > 0 && addRight) {
+                prevWidth = w + prevWidth;
             }
+            w = image.getWidth();
+            final int h = image.getHeight();
+
+            final int[][][] vertexIds = new int[images.size()][prevWidth + w][h];
```
Comment (OrionsGuardian) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:205]: The logic here can be simplified to just use a 2d array:
final int[][] vertexIds = new int[w][h];
Further down in the code, the frame variable is only used in the vertex identifier, so is not needed as part of the array structure. 
Also, the prefWidth should only be used in setting the x-offset the x,y,z co-ordinates of each vertex.
The size of the array should simply match the size of the image.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 7]
```diff
@@ -81,6 +85,9 @@ public class ImageGraphBuilderPlugin extends SimpleEditPlugin {
     private static final Logger LOGGER = Logger.getLogger(ImageGraphBuilderPlugin.class.getName());
 
     public static final String IMAGE_FILE_PARAMETER_ID = PluginParameter.buildId(ImageGraphBuilderPlugin.class, "image_file");
+    public static final String ADD_RIGHT_PARAMETER_ID = PluginParameter.buildId(ImageGraphBuilderPlugin.class, "add_right");
+    public static final String ADD_LAYERS_PARAMETER_ID = PluginParameter.buildId(ImageGraphBuilderPlugin.class, "add_layers");
+    public static final String LAYER_START_PARAMETER_ID = PluginParameter.buildId(ImageGraphBuilderPlugin.class, "layer_start");
```
Comment (OrionsGuardian) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:90]: This doesn't appear to be used anywhere.

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 8]
```diff
@@ -134,127 +155,151 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+
+                    final int yinv = h - y;
+                    ConstructionUtilities.setxyz(graph, vxId, vertexXAttributeId, vertexYAttributeId, vertexZAttributeId, (x + prevWidth) * 2, yinv * 2, -gray * zlen / 255F);
+                    ConstructionUtilities.setxyz(graph, vxId, vertexX2AttributeId, vertexY2AttributeId, vertexZ2AttributeId, (x + prevWidth) * 2, yinv * 2, 0);
+                    graph.setStringValue(vertexBackgroundIconAttributeId, vxId, "Background.Flat Square");
+                    final ConstellationColor color = ConstellationColor.getColorValue(r / 255F, g / 255F, b / 255F, a / 255F);
+                    graph.setObjectValue(vertexColorAttributeId, vxId, color);
+
+                    if (multipleFrames) {
+                        graph.setFloatValue(vertexVisibilityAttributeId, vxId, vis);
```
Comment (OrionsGuardian) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:258]: This seems unnecessary. 
The `vis` value is always 1, which is already the default visibility value.
Should be able to remove the variables and usages of both `vertexVisibilityAttributeId` and `vis`

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java (hunk 7 above)
Comment (andromeda-224) [reply CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java]: Removed LAYER_START_PARAMETER_ID

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java (hunk 8 above)
Comment (andromeda-224) [reply CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java]: Removed vertexVisibilityAttributeId and vis

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 9]
```diff
@@ -134,127 +154,141 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+        int w = 0;
+        int prevWidth = 0;
+        final boolean addRight = (boolean) parameters.getObjectValue(ADD_RIGHT_PARAMETER_ID);
+        final boolean addLayers = (boolean) parameters.getObjectValue(ADD_LAYERS_PARAMETER_ID);
+        final int stateAttributeId = LayersViewConcept.MetaAttribute.LAYERS_VIEW_STATE.ensure(graph);
+        final LayersViewState currentState = graph.getObjectValue(stateAttributeId, 0);
+        if (currentState != null) {
+            // set start layer to existing graph layer count + 1
+            layer = currentState.getLayerCount() + 1;
+        }
```
Comment (antares1470) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:204]: I wonder if this portion of code could only be executed when `addLayers` is true? That way, this attribute is only added and used if there is a need to use it

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 10]
```diff
@@ -134,127 +154,141 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
+                prevWidth = w + prevWidth;
             }
+            w = image.getWidth();
+            final int h = image.getHeight();
+            final int[][] vertexIds = new int[w][h];
+
+            final float zlen = multipleFrames ? 0 : Math.min(w, h) / 4F;
+
+            // get layer bitmask attribute          
+            final int vertexBitmaskAttributeId = LayersConcept.VertexAttribute.LAYER_MASK.ensure(graph);
```
Comment (antares1470) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:225]: similar case here. Probably only want to add this attribute if it is being used (i.e. when `addLayers` is true)

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by antares1470 — APPROVED** [review]

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 11]
```diff
@@ -134,127 +154,141 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
...
-                vertexVisibilityAttributeId = VisualConcept.VertexAttribute.VISIBILITY.get(graph);
-            } else {
-                vertexVisibilityAttributeId = Graph.NOT_FOUND;
+            if (addLayers && layer > 1) {
+                // add new layer for each image if addLayer option selected
+                PluginExecution.withPlugin(new NewLayerPlugin()).executeNow(graph);
+            }
+
+            // If add right option selected, images will be appended to the
+            // right, otherwise, images will be added on top of the prev one
```
[the remark this answers is not in the document]
Comment (andromeda-224) [reply CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java]: It is possible that this functionality may be used to display heatmaps and such. With a few options now available (add images to the right of the previous ones, add images on top of each other, adding each image on separate layers), we may be able to gather more feedback.

File: CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java [hunk 12]
```diff
@@ -247,6 +248,8 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
                     graph.setObjectValue(vertexColorAttributeId, vxId, color);
 
                     if (addLayers) {
+                        // get layer bitmask attribute          
+                        final int vertexBitmaskAttributeId = LayersConcept.VertexAttribute.LAYER_MASK.ensure(graph);
```
Comment (OrionsGuardian) [inline CoreTesting/src/au/gov/asd/tac/constellation/testing/construction/ImageGraphBuilderPlugin.java:252]: This should be defined before entering the nested for loops.
Should not be obtaining/ensuring the attributeId separately for each and every pixel in the image.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Review by OrionsGuardian — APPROVED** [review]

**Pull request merged**